### PR TITLE
Fixed issue #17610: Token does not get marked as used if invalid email address exists

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

File: application/helpers/frontend_helper.php [hunk 1]
```diff
@@ -428,7 +428,9 @@ function submittokens($quotaexit = false)
         }
         $token->usesleft--;
     }
-    $token->save();
+    // We need to specify the attributes we are updating because validation could fail on other (untouched) attributes,
+    // preventing the update (eg.: the participant could have an invalid email).
+    $token->save(true, ['completed', 'usesleft']);
```
Comment (Shnoulle) [inline application/helpers/frontend_helper.php:433]: :+1:  maybe https://www.yiiframework.com/doc/api/1.1/CActiveRecord#saveAttributes-detail is better ?

Comment (olleharstedt) [reply application/helpers/frontend_helper.php:433]: saveAttributes does not run validation, it seems.

Comment (Shnoulle) [reply application/helpers/frontend_helper.php:433]: Here : we need to don't validate …  we need to disable validation …
We validate only `['completed', 'usesleft']` …

Comment (olleharstedt) [reply application/helpers/frontend_helper.php:433]: >  we need to disable validation …

Yes, and better to be explicit about it. Yii 1 is an absolute mess in this regard already. save() validates but update() doesn't? Pfff.

Comment (Shnoulle) [reply application/helpers/frontend_helper.php:433]: I mean `$token->save(true, ['completed', 'usesleft']);` : https://www.yiiframework.com/doc/api/1.1/CActiveRecord#save-detail i reamlly unclear about validation : true mean validate, but validate what ?

 `$token->save(false)` is clear : no validation. 

`$token->saveAttributes(['completed', 'usesleft']);` seems really clear (for me) : just save the attributes in the array: no validation is done :)

Comment (gabrieljenik) [reply application/helpers/frontend_helper.php:433]: `$token->saveAttributes(['completed', 'usesleft'])` doesn't perform validation on those attributes.

`$token->save(true, ['completed', 'usesleft']);` only valdates and saves those attributes.
What's wrong about that?

**Pull request merged**